replica/table.cc

```diff
@@ -446,25 +447,48 @@ void compaction_group::backlog_tracker_adjust_charges(const std::vector<sstables
     tracker.replace_sstables(old_sstables, new_sstables);
 }
 
-lw_shared_ptr<sstables::sstable_set>
-compaction_group::do_add_sstable(lw_shared_ptr<sstables::sstable_set> sstables, sstables::shared_sstable sstable,
-        enable_backlog_tracker backlog_tracker) {
-    if (belongs_to_other_shard(sstable->get_shards_for_this_sstable())) {
-        on_internal_error(tlogger, format("Attempted to load the shared SSTable {} at table", sstable->get_filename()));
+compaction_group_sstables_adder::compaction_group_sstables_adder(compaction_group& cg, std::vector<sstables::shared_sstable> sstables, is_main main)
+    : _cg(cg)
+    , _sstables(std::move(sstables))
+    , _main(main)
+{
+    for (const auto& sst : sstables) {
+        if (belongs_to_other_shard(sst->get_shards_for_this_sstable())) {
+            on_internal_error(tlogger, format("Attempted to load the shared SSTable {} at table", sst->get_filename()));
+        }
     }
+}
+
+// Guarantees strong exception safety
+future<> compaction_group_sstables_adder::prepare() {
     // allow in-progress reads to continue using old list
-    auto new_sstables = make_lw_shared<sstables::sstable_set>(*sstables);
-    new_sstables->insert(sstable);
-    if (backlog_tracker) {
-        table::add_sstable_to_backlog_tracker(get_backlog_tracker(), sstable);
+    auto cur = _main ? _cg.main_sstables() : _cg.maintenance_sstables();
+    _new_sstable_set = make_lw_shared<sstables::sstable_set>(*cur);
+    _new_sstable_set->insert(_sstables);
+    return make_ready_future<>();
+}
+
+void compaction_group_sstables_adder::execute() {
+    if (_main) {
+        _cg._main_sstables = std::move(_new_sstable_set);
+        for (const auto& sst : _sstables) {
+            _cg._main_set_disk_space_used += sst->bytes_on_disk();
```

Comment:
Perhaps we should make sstable_set track this (not now).

Reply:
Yes, it's doable, although an optimization.